x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java

```diff
@@ -177,19 +181,64 @@ private void onFatalFailure(AsyncSearchTask task, Exception error, boolean shoul
     private void onFinalResponse(AsyncSearchTask searchTask,
                                  AsyncSearchResponse response,
                                  Runnable nextAction) {
-        store.updateResponse(searchTask.getExecutionId().getDocId(), threadContext.getResponseHeaders(),response,
-            ActionListener.wrap(resp -> unregisterTaskAndMoveOn(searchTask, nextAction),
+        store.updateResponse(searchTask.getExecutionId().getDocId(), threadContext.getResponseHeaders(), response,
+            ActionListener.wrap(
+                resp -> unregisterTaskAndMoveOn(searchTask, nextAction),
                 exc -> {
                     Throwable cause = ExceptionsHelper.unwrapCause(exc);
                     if (cause instanceof DocumentMissingException == false &&
                             cause instanceof VersionConflictEngineException == false) {
                         logger.error(() -> new ParameterizedMessage("failed to store async-search [{}]",
                             searchTask.getExecutionId().getEncoded()), exc);
+                        updateStoredResponseWithFailure(
```

Comment:
I wonder if we should create a failure response inside `updateResponse` method instead? I worry that future usages might forget creating corresponding failures. I think we can do it by adding a new method that returns a failure response to AsyncResponse. WDYT?

Reply:
Good point!

Reply:
Addressed in d03015c